query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+      )
+    ) {
+      throw new GraphQLError(
+        `Unsupported core schema feature and/or version: ${feature}`,
+        schema.astNode!,
+      );
+    }
+  }
+
+  const joinName = 'join';
+
+  function getJoinDirective(name: string) {
+    const fullyQualifiedName = `${joinName}__${name}`;
+
+    const directive = schema.getDirective(fullyQualifiedName);
+    assert(
+      directive,
+      `Composed schema should define @${fullyQualifiedName} directive`,
+    );
+    return directive;
+  }
+
+  const ownerDirective = getJoinDirective('owner');
+  const typeDirective = getJoinDirective('type');
+  const fieldDirective = getJoinDirective('field');
+  const graphDirective = getJoinDirective('graph');
+
+  const graphEnumType = schema.getType(`${joinName}__Graph`);
+  assert(isEnumType(graphEnumType));
+
+  const graphMap: GraphMap = Object.create(null);
+
+  schema.extensions = {
+    ...schema.extensions,
+    federation: {
+      graphs: graphMap,
+    },
+  };
+
+  for (const graphValue of graphEnumType.getValues()) {
+    const name = graphValue.name;
+
+    const graphDirectiveArgs = getArgumentValuesForDirective(
+      graphDirective,
+      graphValue.astNode!,
+    );
+    assert(
+      graphDirectiveArgs,
+      `${graphEnumType.name} value ${name} in composed schema should have a @${graphDirective.name} directive`,
+    );
+
+    const serviceName: string = graphDirectiveArgs['name'];
+    const url: string = graphDirectiveArgs['url'];
+
+    graphMap[name] = {
+      serviceName,
+      url,
+    };
+  }
+
+  for (const type of Object.values(schema.getTypeMap())) {
+    if (isIntrospectionType(type)) continue;
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
+      ownerDirective,
+      type.astNode,
+    );
+
+    const typeMetadata: FederationTypeMetadata = ownerDirectiveArgs
+      ? {
+          serviceName: graphMap[ownerDirectiveArgs?.['graph']].serviceName,
+          keys: new MultiMap(),
+          isValueType: false,
+        }
+      : {
+          isValueType: true,
+        };
+
+    type.extensions = {
+      ...type.extensions,
+      federation: typeMetadata,
+    };
+
+    const typeDirectivesArgs = getArgumentValuesForRepeatableDirective(
+      typeDirective,
+      type.astNode,
+    );
+
+    assert(
+      !(typeMetadata.isValueType && typeDirectivesArgs.length >= 1),
+      `GraphQL type "${type.name}" cannot have a @${typeDirective.name} \
+directive without an $${ownerDirective.name} directive`,
+    );
+
+    for (const typeDirectiveArgs of typeDirectivesArgs) {
+      const serviceName = graphMap[typeDirectiveArgs['graph']].serviceName;
+
+      const keyFields = parseFieldSet(typeDirectiveArgs['key']);
+
+      typeMetadata.keys?.add(serviceName, keyFields);
```

Comment:
So if this type lacks `@join__owner` but has `@join__type`, we... validate that the `graph` is known and that the `key` parses but then do nothing?  This seems reasonable but if it's an explicit choice add a comment noting that it's intentional (and that it wouldn't be a good optimization to put the whole `for` body into `if (typeMetadata.keys) {}`)